pycbc/waveform/generator.py

```diff
@@ -1225,30 +1229,57 @@ def select_waveform_generator(approximant):
+            'cbc': FDomainCBCGenerator,
+            'ringdown': {
+                'FdQNMfromFinalMassSpin': FDomainMassSpinRingdownGenerator,
+                'FdQNMfromFreqTau': FDomainFreqTauRingdownGenerator,
+            }
+        },
+        'td': {
+            'cbc': TDomainCBCGenerator,
+            'ringdown': {
+                'TdQNMfromFinalMassSpin': TDomainMassSpinRingdownGenerator,
+                'TdQNMfromFreqTau': TDomainFreqTauRingdownGenerator,
+            },
+            'supernovae': {
+                'CoreCollapseBounce': TDomainSupernovaeGenerator
+            }
+        }
+    }
+
+    gen = None
+    if domain not in {None, 'td', 'fd'}:
+        raise ValueError(f"Invalid domain '{domain}'. Must be one of: None, 'td', or 'fd'.")
+
+    # Check if the approximant exists in any domain
     if approximant in waveform.fd_approximants():
-        return FDomainCBCGenerator
-    # check if time-domain CBC waveform
+        if domain == 'fd' or domain is None:
+            gen = generator_map['fd']['cbc']
+        elif domain == 'td' and approximant in waveform.td_approximants():
+            gen = generator_map['td']['cbc']
+        elif domain == 'td':
+            raise ValueError(f"Requested domain is 'td', but {approximant} is a frequency-domain model.")
     elif approximant in waveform.td_approximants():
-        return TDomainCBCGenerator
-    # check if frequency-domain ringdown waveform
+        if domain == 'td' or domain is None:
+            gen = generator_map['td']['cbc']
+        elif domain == 'fd' and approximant in waveform.fd_approximants():
+            gen = generator_map['fd']['cbc']
+        elif domain == 'fd':
+            raise ValueError(f"Requested domain is 'fd', but {approximant} is a time-domain model.")
     elif approximant in ringdown.ringdown_fd_approximants:
-        if approximant == 'FdQNMfromFinalMassSpin':
-            return FDomainMassSpinRingdownGenerator
-        elif approximant == 'FdQNMfromFreqTau':
-            return FDomainFreqTauRingdownGenerator
+        gen = generator_map['fd']['ringdown'].get(approximant, None)
     elif approximant in ringdown.ringdown_td_approximants:
-        if approximant == 'TdQNMfromFinalMassSpin':
-            return TDomainMassSpinRingdownGenerator
-        elif approximant == 'TdQNMfromFreqTau':
-            return TDomainFreqTauRingdownGenerator
-    # check if supernovae waveform:
+        gen = generator_map['td']['ringdown'].get(approximant, None)
     elif approximant in supernovae.supernovae_td_approximants:
-        if approximant == 'CoreCollapseBounce':
-            return TDomainSupernovaeGenerator
-    # otherwise waveform approximant is not supported
-    else:
-        raise ValueError("%s is not a valid approximant." % approximant)
+        gen = generator_map['td']['supernovae'].get(approximant, None)
+
+    if gen is None:
+        raise ValueError(f"{approximant} is not a valid approximant.")
+
+    return gen
```

Comment:
I kind of agree with code climate here... the logic is challenging to follow. More importantly, it's not good that domain is silently ignored if you are using a ringdown or supernova approximants. I think this could be simplified and made more uniform if you make the initial decision based on domain rather than approximant. Something like:
```
if domain == 'td':
    gen = get_td_generator(approximant)
elif domain == 'fd':
    gen = get_fd_generator(approximant)
elif domain is None:
    try:
        gen = get_fd_generator(approximant)
    except ValueError:
         gen = get_td_generator(approximant)
```
Then write separate functions `get_td_generator` and `get_fd_generator` that will pull out the appropriate cbc, ringdown, or supernova generator based on the approximant, raising a ValueError if it can't find one. That will also be more straight forward to explain: if the domain is none, it will first try the FD approximants, then try the TD approximants, then fail if it can't find either, and would apply equally to all types of signals.

Doing this means you'll have to move the generator_map out of the function definition, since get_td_generator and get_fd_generator will need to access it. Better yet, since you'll have separate functions for this, you can drop the dictionary and just have each function apply the appropriate logic to select the generator.

Reply:
Thanks for the feedback, I agree. I initially had the code logic set up in this way. If I revert to it codeclimate might set up a flag for having 20 branches. If it is alright, I can revert these changes